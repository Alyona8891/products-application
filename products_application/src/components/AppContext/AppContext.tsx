import { createContext, useCallback, useState } from 'react';
import { AppContextProps, IRequestResult } from '../../types/types';
import { getProducts } from '../../utils/api';
import { DEFAULT_ITEMS_QUANTITY } from '../../constants/constants';

export const AppContext = createContext({} as AppContextProps);

export function ProductsProvider({
  children,
}: {
  children: React.ReactNode;
}): React.ReactElement {
  const [productsData, setProductsData] = useState<IRequestResult>({
    products: [],
    total: 0,
  });
  const [quantityProductsOnPage, setProductsOnPage] = useState(
    DEFAULT_ITEMS_QUANTITY
  );
  const [isLoadingProducts, setIsLoadingProducts] = useState(true);
  const [isLoadingPagination, setIsLoadingPagination] = useState(true);
  const [inputValue, setInputValue] = useState('');

  const getProductsData = useCallback(
    (keyWord: string, currentPage: number, quantityProductsOnPage: number) => {
      getProducts(keyWord, currentPage, quantityProductsOnPage).then((data) => {
        if (data) {
          setProductsData({
            products: data.products,
            total: data.total,
          });
        }
        setIsLoadingPagination(false);
        setIsLoadingProducts(false);
      });
    },
    []
  );

  return (
    <AppContext.Provider
      value={{
        productsData,
        isLoadingProducts,
        isLoadingPagination,
        getProductsData,
        setIsLoadingProducts,
        inputValue,
        setInputValue,
        setProductsOnPage,
        quantityProductsOnPage,
      }}
    >
      {children}
    </AppContext.Provider>
  );
}
