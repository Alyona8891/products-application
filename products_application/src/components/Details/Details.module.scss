.details {
    position: fixed;
    background-color: initial;
    width: 100%;
    height: 100%;
    top: 0;
    right: 0;
}

.shadow {
    position: fixed;
    background-color: rgba(0, 0, 0, 0.2);
    width: 100%;
    height: 100%;
    top: 0;
    right: 0;
    transition: background-color 0.3s ease-in-out;
}

.container {
    position: absolute;
    top: 0;
    right: 0;
    background-color: yellowgreen;
    width: 20%;
    height: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 10px;
    justify-content: center;
    animation: slidein 1s ease-in-out;
    transform: translateX(0);
    z-index: 2;
}

@keyframes slidein {
    from {
        transform: translateX(100%);
    }
  
    to {
        transform: translateX(0);
    }
  }

.close_button {
    background-color: inherit;
    position: absolute;
    top: 10px;
    right: 10px;
    border: 0;
    cursor: pointer;
    transform: rotate(45deg);
    font-size: 30px;
    z-index: 3;
    transition: 0.3s;
    &:hover {
        color: rgb(148, 13, 13);
        transition: 0.3s;
    }
}

.image {
    width: 19vw;
    margin-bottom: 10px;
    max-height: 40vh;
}

.title {
    font-size: calc(10px + (26 - 20) * ((100vw - 320px) / (1920 - 320)));
    text-align: left;
    background-color: inherit;
    text-transform: uppercase;
    margin-bottom: 10px;
}

.text {
    align-self: end;
    height: 150px;
    text-align: justify;
    background-color: inherit;
    border-bottom-right-radius: 1.5rem;
    border-bottom-left-radius: 1.5rem;
    font-size: calc(10px + (26 - 20) * ((100vw - 320px) / (1920 - 320)));
    max-width: 100%;
    word-wrap: break-word;
}

.message {
    position: absolute;
    top: 45vh;
    display: flex;
    align-items: center;
    justify-content: center;
    font-weight: bold;
    font-size: 30px;
    font-family: monospace;
    width: 100%;
    text-align: center;
    margin: 0 auto;
    background-color: inherit;
}