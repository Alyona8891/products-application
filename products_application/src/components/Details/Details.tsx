import styles from './Details.module.scss';
import { IProduct } from '../../types/types';

export function Details(props: {
  data: IProduct;
  handleQueryChange: (
    search?: string,
    page?: number,
    productsOnPage?: number,
    details?: number
  ) => void;
}): React.ReactElement {
  const { data, handleQueryChange } = props;

  return (
    <section className={styles.details} data-testid="details">
      <div
        className={styles.shadow}
        onClick={() => handleQueryChange(undefined, undefined, undefined, 0)}
      />
      {data?.title ? (
        <div className={styles.container}>
          <button
            onClick={() =>
              handleQueryChange(undefined, undefined, undefined, 0)
            }
            className={styles.close_button}
            data-testid="closeButton"
          >
            +
          </button>
          <img
            className={styles.image}
            src={data.images[0]}
            alt="detail image"
          />
          <h3 className={styles.title}>{data.title}</h3>
          <p className={styles.text}>{data.description}</p>
        </div>
      ) : (
        <div className={styles.container}>
          <button
            onClick={() =>
              handleQueryChange(undefined, undefined, undefined, 0)
            }
            className={styles.close_button}
            data-testid="closeButton"
          >
            +
          </button>
          <p className={styles.message}>
            Sorry, nothing found. Please, try again!
          </p>
        </div>
      )}
    </section>
  );
}
