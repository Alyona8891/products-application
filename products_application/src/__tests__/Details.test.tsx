import { cleanup, render, screen } from '@testing-library/react';
import { AppContext } from '../components/AppContext/AppContext';
import { afterEach, expect, test } from 'vitest';
import { MemoryRouter } from 'react-router-dom';
import { Details } from '../components/Details/Details';

afterEach(() => cleanup());

test('displays a loading indicator is displayed while fetching data', async () => {
  const mockContext = {
    productsData: {
      products: [
        {
          id: 1,
          title: 'ijivjri',
          text: 'okcovo',
          images: ['ofo'],
          description: 'okcovo',
        },
        {
          id: 2,
          title: 'efe',
          text: 'efef',
          images: ['efefe'],
          description: 'efefe',
        },
        {
          id: 3,
          title: 'efe',
          text: 'efef',
          images: ['efefe'],
          description: 'efefe',
        },
      ],
      total: 1,
    },
    isLoadingProducts: false,
    isLoadingPagination: false,
    getProductsData: () => {},
    setIsLoadingProducts: () => {},
    inputValue: '',
    setInputValue: () => {},
    quantityProductsOnPage: 10,
    setProductsOnPage: () => {},
  };
  render(
    <AppContext.Provider value={mockContext}>
      <MemoryRouter>
        <Details />
      </MemoryRouter>
    </AppContext.Provider>
  );
  await screen.findByText('Loading');
  const elements = screen.getByText('Loading');
  expect(elements).toBeInTheDocument();
});

/*test('detailed card component displays the correct product information', async () => {
  render(<App />);
  await waitFor(() => {
    const card = screen.getAllByTestId('card')[0];
    fireEvent.click(card);
    expect(location.search).toBe('?page=1&details=1');
    const details = screen.findByTestId('details');
    expect(details).toMatchSnapshot();
    expect(screen.getByTestId('details')).toBeInTheDocument();
  });
});*/
