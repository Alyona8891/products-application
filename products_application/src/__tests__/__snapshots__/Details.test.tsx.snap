// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`detailed card component displays the correct product information 1`] = `
<div
  class="_container_0c9c7f"
  data-testid="details"
>
  <a
    class="_close_button_0c9c7f"
    href="/?page=1"
  >
    +
  </a>
  <img
    alt="detail image"
    class="_image_0c9c7f"
    src="https://i.dummyjson.com/data/products/1/1.jpg"
  />
  <h3
    class="_title_0c9c7f"
  >
    iPhone 9
  </h3>
  <p
    class="_text_0c9c7f"
  >
    An apple mobile which is nothing like apple
  </p>
</div>
`;

exports[`detailed card component displays the correct product information 2`] = `Promise {}`;

exports[`detailed card component displays the correct product information 3`] = `Promise {}`;
